File: apps/vet-windows/src/TimiVet/Services/ClinicApiClient.cs
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using TimiVet.Models;

namespace TimiVet.Services;

/// <summary>
/// A failure talking to the Tími Worker, carrying the HTTP status when there was one.
///
/// The status is what lets the console tell "the Worker refused this credential" apart from "there is no
/// network right now". Both used to arrive as an <see cref="InvalidOperationException"/> with a sentence
/// in it, so the poll loop had no way to decide between reconnecting quietly and asking somebody to sign
/// in — and it always chose one of them wrongly.
/// </summary>
public sealed class ClinicApiException(string message, HttpStatusCode? statusCode = null) : Exception(message)
{
    public HttpStatusCode? StatusCode { get; } = statusCode;

    /// <summary>The Worker rejected who we are, rather than failing to answer.</summary>
    public bool IsAuthenticationFailure => StatusCode is HttpStatusCode.Unauthorized or HttpStatusCode.Forbidden;

    /// <summary>Nothing answered at all — DNS, TLS, a timeout, a captive portal. Retry, do not sign out.</summary>
    public bool IsUnreachable => StatusCode is null;
}

public sealed class ClinicApiClient
{
    private static readonly JsonSerializerOptions Json = new(JsonSerializerDefaults.Web) { PropertyNameCaseInsensitive = true };
    private readonly HttpClient _http = new() { Timeout = TimeSpan.FromSeconds(20) };
    private readonly DemoClinicData _demo = new();
    private readonly ClerkAuthService _auth;
    private AppSettings _settings;

    public ClinicApiClient(AppSettings settings, ClerkAuthService auth)
    {
        _settings = settings;
        _auth = auth;
    }

    public bool IsDemo => !Uri.TryCreate(_settings.ApiBaseUrl, UriKind.Absolute, out var uri) || uri.Scheme != Uri.UriSchemeHttps;
    public void UpdateSettings(AppSettings settings) => _settings = settings;

    /// <summary>The address actually in use, so an error can quote it. A blank one is the most common
    /// cause of "could not reach Clerk" and the least visible thing on the screen.</summary>
    public string ConfiguredAddress =>
        string.IsNullOrWhiteSpace(_settings.ApiBaseUrl) ? "no Worker address" : _settings.ApiBaseUrl.Trim();

    public async Task<ClinicDashboard> GetDashboardAsync(CancellationToken cancellationToken)
    {
        if (IsDemo) return _demo.Dashboard();
        using var response = await SendAsync(HttpMethod.Get, "/api/clinic/dashboard", null, cancellationToken);
        return await ReadAsync<ClinicDashboard>(response, cancellationToken);
    }

    /// <summary>
    /// GET /api/clinic/payouts — what Tími has sent this clinic and what Stripe has paid on.
    /// </summary>
    /// <remarks>
    /// Answers an empty set rather than throwing when payments are not configured. A practice that has
    /// not finished Stripe onboarding still opens this panel, and an error where three zeroes belong
    /// reads as a broken console rather than as an account that is not set up yet.
    /// </remarks>
    public async Task<ClinicPayouts> GetPayoutsAsync(CancellationToken cancellationToken)
    {
        if (IsDemo) return new ClinicPayouts();
        using var response = await SendAsync(HttpMethod.Get, "/api/clinic/payouts", null, cancellationToken);
        return await ReadAsync<ClinicPayouts>(response, cancellationToken);
    }

    /// <summary>
    /// GET /api/config — the Clerk publishable key, and therefore the first request the console ever
    /// makes. Answered to anyone by the Worker, and sent without a session by <see cref="CreateAsync"/>.
    /// </summary>
    public async Task<AppConfig> GetConfigAsync(CancellationToken cancellationToken)
    {
        using var response = await SendAsync(HttpMethod.Get, "/api/config", null, cancellationToken);
        return await ReadAsync<AppConfig>(response, cancellationToken);
    }

    public async Task<SessionDescriptor> GetSessionAsync(CancellationToken cancellationToken)
    {
        using var response = await SendAsync(HttpMethod.Get, "/api/session", null, cancellationToken);
        var envelope = await ReadAsync<SessionEnvelope>(response, cancellationToken);
        return envelope.Session;
    }

    public async Task<TenantMembersResponse> GetTenantMembersAsync(CancellationToken cancellationToken)
    {
        using var response = await SendAsync(HttpMethod.Get, "/api/tenant/members", null, cancellationToken);
        return await ReadAsync<TenantMembersResponse>(response, cancellationToken);
    }

    public Task AddTenantMemberAsync(string email, string role, CancellationToken cancellationToken) =>
        SendAndDiscardAsync(HttpMethod.Post, "/api/tenant/members", new { email, role }, cancellationToken);

    public Task ChangeTenantMemberRoleAsync(string userId, string role, CancellationToken cancellationToken) =>
        SendAndDiscardAsync(HttpMethod.Patch, $"/api/tenant/members/{Uri.EscapeDataString(userId)}", new { role }, cancellationToken);

    public Task RemoveTenantMemberAsync(string userId, CancellationToken cancellationToken) =>
        SendAndDiscardAsync(HttpMethod.Delete, $"/api/tenant/members/{Uri.EscapeDataString(userId)}", null, cancellationToken);

    public Task RevokeInvitationAsync(string invitationId, CancellationToken cancellationToken) =>
        SendAndDiscardAsync(HttpMethod.Delete, $"/api/tenant/invitations/{Uri.EscapeDataString(invitationId)}", null, cancellationToken);

    // ---- Calling preferences (GET/PATCH /api/clinic/call-preferences) ----

    public async Task<CallPreferences> GetCallPreferencesAsync(CancellationToken cancellationToken)
    {
        if (IsDemo) return new CallPreferences { CallPolicy = "always", CallsEnabled = true, LocationPhone = "(510) 555-0148" };
        using var response = await SendAsync(HttpMethod.Get, "/api/clinic/call-preferences", null, cancellationToken);
        var envelope = await ReadAsync<CallPreferencesEnvelope>(response, cancellationToken);
        return envelope.Preferences;
    }

    /// <summary>
    /// PATCH, not POST, and only the fields being changed — see <see cref="CallPreferencesUpdate"/> for
    /// why a null that reaches the wire is a setting rather than an omission. Reading these is open to
    /// everybody on the team; writing them is an administrator's call, and the Worker enforces that.
    /// </summary>
    public async Task<CallPreferences> UpdateCallPreferencesAsync(CallPreferencesUpdate update, CancellationToken cancellationToken)
    {
        if (IsDemo)
        {
            var policy = update.CallPolicy ?? "always";
            return new CallPreferences
            {
                CallPolicy = policy,
                CallsEnabled = policy != "never",
                VoicePhone = string.IsNullOrWhiteSpace(update.VoicePhone) ? null : update.VoicePhone,
                LocationPhone = "(510) 555-0148",
                QuietHours = update.QuietHours
            };
        }
        using var response = await SendAsync(HttpMethod.Patch, "/api/clinic/call-preferences", update, cancellationToken);
        var envelope = await ReadAsync<CallPreferencesEnvelope>(response, cancellationToken);
        return envelope.Preferences;
    }

    public async Task PublishAvailabilityAsync(AvailabilityUpdate update, CancellationToken cancellationToken)
    {
        if (IsDemo) { _demo.Publish(update); return; }
        using var response = await SendAsync(HttpMethod.Post, "/api/clinic/availability", update, cancellationToken);
        await EnsureSuccessAsync(response, cancellationToken);
    }

    public async Task RespondAsync(ClinicRequest item, ClinicDecision decision, CancellationToken cancellationToken)
    {
        if (IsDemo) { _demo.Decide(item.Id, decision); return; }
        object payload = item.SearchTarget
            ? new { decision = decision.Decision, responseType = decision.ResponseType, availableAt = decision.AvailableAt, arrivalWindowMinutes = decision.ArrivalWindowMinutes, holdMinutes = decision.HoldMinutes, waitMin = decision.WaitMin, waitMax = decision.WaitMax, note = decision.Note }
            : new { decision = decision.Decision == "offer" ? "accept" : "decline", arrivalWindowMinutes = decision.ArrivalWindowMinutes, note = decision.Note };
        var path = item.SearchTarget ? $"/api/clinic/search-targets/{Uri.EscapeDataString(item.Id)}/decision" : $"/api/clinic/intakes/{Uri.EscapeDataString(item.Id)}/decision";
        using var response = await SendAsync(HttpMethod.Post, path, payload, cancellationToken);
        await EnsureSuccessAsync(response, cancellationToken);
    }

    // ---- Analytics (POST /api/analytics) --------------------------------

    /// <summary>
    /// Fire-and-forget beacon to POST /api/analytics.
    /// </summary>
    /// <remarks>
    /// Deliberately not routed through <see cref="CreateAsync"/>: the endpoint is public and cookieless,
    /// so the beacon must carry no Worker token, demo header, or any other identifier. It is never
    /// awaited and every failure is swallowed whole — a metrics beacon that can surface an error, or
    /// hold up an intake decision, in a clinic's console has its priorities backwards.
    /// </remarks>
    public void TrackEvent(string name, Dictionary<string, string>? meta = null)
    {
        if (IsDemo || string.IsNullOrWhiteSpace(_settings.ApiBaseUrl)) return;
        var url = new Uri(new Uri(_settings.ApiBaseUrl.Trim().TrimEnd('/') + "/"), "api/analytics");
        _ = Task.Run(async () =>
        {
            try
            {
                using var request = new HttpRequestMessage(HttpMethod.Post, url)
                {
                    Content = JsonContent.Create(new AnalyticsPayload { Events = [new AnalyticsEvent { Name = name, Meta = meta }] }, options: Json)
                };
                using var response = await _http.SendAsync(request, CancellationToken.None);
            }
            catch
            {
                // Silent by contract: a failed beacon must never reach a screen, a toast, or a log line
                // somebody at a front desk might read as the console being broken.
            }
        });
    }

    /// <summary>The console-analytics wire shape: {events: [{name, path?, meta?}]}, at most 25 events,
    /// and — because the endpoint is cookieless by design — nothing that identifies the operator or the
    /// tenant.</summary>
    private sealed class AnalyticsPayload
    {
        public List<AnalyticsEvent> Events { get; set; } = [];
    }

    private sealed class AnalyticsEvent
    {
        public string Name { get; set; } = "";
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public Dictionary<string, string>? Meta { get; set; }
    }

    private async Task SendAndDiscardAsync(HttpMethod method, string path, object? body, CancellationToken cancellationToken)
    {
        using var response = await SendAsync(method, path, body, cancellationToken);
        await EnsureSuccessAsync(response, cancellationToken);
    }

    /// <summary>
    /// Sends one request, re-minting the Worker token and retrying exactly once on a 401. The retry goes
    /// through <see cref="CreateAsync"/> again rather than reusing the first message, so it genuinely
    /// carries the newly minted token instead of replaying the dead one that caused the 401.
    ///
    /// Demo headers (x-demo-role / x-demo-tenant-id) are only ever attached when there is no real Clerk
    /// session AND the resolved host is a localhost/loopback development address — never against a
    /// production HTTPS Worker.
    /// </summary>
    private async Task<HttpResponseMessage> SendAsync(HttpMethod method, string path, object? body, CancellationToken cancellationToken)
    {
        var response = await SendOnceAsync(method, path, body, cancellationToken);
        if (response.StatusCode == HttpStatusCode.Unauthorized && _auth.HasSession)
        {
            response.Dispose();
            await _auth.ForceRefreshAsync(cancellationToken);
            response = await SendOnceAsync(method, path, body, cancellationToken);
        }
        return response;
    }

    /// <summary>
    /// One attempt, with transport failures translated into a <see cref="ClinicApiException"/> carrying no
    /// status. That absence is load-bearing: it is how the poll loop knows the difference between the
    /// Worker saying no and nothing having answered at all.
    /// </summary>
    private async Task<HttpResponseMessage> SendOnceAsync(HttpMethod method, string path, object? body, CancellationToken cancellationToken)
    {
        try
        {
            using var request = await CreateAsync(method, path, body, cancellationToken);
            return await _http.SendAsync(request, cancellationToken);
        }
        catch (HttpRequestException ex)
        {
            throw new ClinicApiException($"Could not reach {ConfiguredAddress} — {ex.Message}");
        }
        catch (TaskCanceledException) when (!cancellationToken.IsCancellationRequested)
        {
            throw new ClinicApiException($"{ConfiguredAddress} did not answer within 20 seconds.");
        }
        catch (ClerkApiException ex)
        {
            // Minting the Worker token is part of sending the request, so a Clerk refusal has to arrive
            // wearing the same clothes as a Worker refusal — otherwise the caller sees an exception type
            // it does not recognise, files it under weather, and reconnects forever against a session
            // that has genuinely been revoked.
            throw new ClinicApiException(ex.Message, ex.IsCredentialRejected ? HttpStatusCode.Unauthorized : null);
        }
    }

    private async Task<HttpRequestMessage> CreateAsync(HttpMethod method, string path, object? body, CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(_settings.ApiBaseUrl))
            throw new ClinicApiException("Enter the HTTPS address of your Tími Worker in Settings.");
        var url = new Uri(new Uri(_settings.ApiBaseUrl.Trim().TrimEnd('/') + "/"), path.TrimStart('/'));
        var request = new HttpRequestMessage(method, url);
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

        if (_auth.HasSession)
        {
            await _auth.EnsureFreshTokenAsync(cancellationToken);
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _auth.WorkerToken);
        }
        else if (IsLoopback(url))
        {
            request.Headers.Add("x-demo-role", "clinic");
            request.Headers.Add("x-demo-tenant-id", _settings.TenantId);
        }
        else if (!IsPublic(url))
        {
            request.Dispose();
            throw new ClinicApiException("Sign in to Tími before contacting a production Worker.");
        }

        if (body is not null) request.Content = JsonContent.Create(body, options: Json);
        return request;
    }

    /// <summary>
    /// The endpoints the Worker answers to anyone, and that the console must reach <em>before</em> it can
    /// sign in.
    ///
    /// /api/config is where the Clerk publishable key comes from, so demanding a session in order to fetch
    /// it is a deadlock: no config, no Clerk host, no sign-in, no session, no config. The macOS console
    /// reported it as "Could not read https://providers.timinow.pet/api/config — Sign in to Tími before
    /// contacting a production Worker", which reads as a Worker or a Clerk problem and is neither; the
    /// request was never sent. This client only escaped it because ClerkAuthService used to keep a private
    /// HttpClient to fetch the config behind the gate's back. That back door is gone, so the exemption
    /// here is what keeps sign-in possible at all.
    /// </summary>
    private static bool IsPublic(Uri url) => url.AbsolutePath.TrimEnd('/').EndsWith("/api/config", StringComparison.OrdinalIgnoreCase);

    private static bool IsLoopback(Uri url) => url.IsLoopback || url.Host.Equals("localhost", StringComparison.OrdinalIgnoreCase);

    private static async Task<T> ReadAsync<T>(HttpResponseMessage response, CancellationToken token)
    {
        await EnsureSuccessAsync(response, token);
        try
        {
            return await response.Content.ReadFromJsonAsync<T>(Json, token)
                ?? throw new ClinicApiException("The Tími API returned an empty response.", response.StatusCode);
        }
        catch (JsonException)
        {
            throw new ClinicApiException("The Tími API returned a response Tími could not read.", response.StatusCode);
        }
    }

    private static async Task EnsureSuccessAsync(HttpResponseMessage response, CancellationToken token)
    {
        if (response.IsSuccessStatusCode) return;
        var raw = await response.Content.ReadAsStringAsync(token);
        string? message = null;
        try
        {
            using var document = JsonDocument.Parse(string.IsNullOrWhiteSpace(raw) ? "{}" : raw);
            var root = document.RootElement;
            if (root.ValueKind == JsonValueKind.Object)
            {
                if (root.TryGetProperty("message", out var value) && value.ValueKind == JsonValueKind.String) message = value.GetString();
                else if (root.TryGetProperty("error", out var error) && error.ValueKind == JsonValueKind.String) message = error.GetString();
            }
        }
        catch (JsonException)
        {
            // An HTML error page from a proxy is not the Worker speaking. Fall through to the status.
        }
        throw new ClinicApiException(
            string.IsNullOrWhiteSpace(message) ? $"Tími API error {(int)response.StatusCode}." : message!,
            response.StatusCode);
    }
}

File: apps/vet-desktop/Sources/TimiVetCore/ClinicAPIClient.swift
import Foundation

// Swift port of apps/vet-windows/src/TimiVet/Services/ClinicApiClient.cs.
//
// Skip rules honored here (see scripts/validate-native.mjs for the customer
// app's equivalent checks):
//   - URLSession never appears in a public initializer signature.
//   - ClinicAPIError is a plain `Error` enum with a `message` payload; it does
//     NOT adopt `LocalizedError` or define `errorDescription`, because Skip
//     cannot translate that Swift-only conformance to Kotlin.

public enum ClinicAPIError: Error, Sendable {
    case invalidConfiguration
    case signInRequired
    case server(String)
    case invalidResponse

    public var message: String {
        switch self {
        case .invalidConfiguration: return "Enter the HTTPS address of your Tími Worker in Settings."
        case .signInRequired: return "Sign in to Tími before contacting a production Worker."
        case .server(let message): return message
        case .invalidResponse: return "The Tími API returned a response it could not read."
        }
    }
}

/// Lets `ClinicAPIClient` attach a live Clerk session token without holding a
/// strong (or even a type-level) dependency on `AuthController` — avoids a
/// retain cycle and keeps the two types independently testable. AuthController
/// conforms to this in the same file it is declared.
public protocol ClinicSessionTokenProviding: AnyObject, Sendable {
    /// `async` because the real conformer (`AuthController`) is
    /// `@MainActor`-isolated; a plain synchronous requirement could not be
    /// satisfied by an actor-isolated computed property.
    var hasSession: Bool { get async }
    /// Returns the current token, minting or refreshing it first only if it is
    /// missing or near expiry.
    func ensureFreshToken() async throws -> String
    /// Unconditionally mints a new token — used once after a 401.
    func forceRefreshToken() async throws -> String
}

/// Talks to the veterinary console API (`/api/clinic/*`, `/api/session`,
/// `/api/config`, `/api/tenant/members`) or, when no HTTPS base URL is
/// configured, falls back to `DemoClinicData` exactly like the Windows client.
public final class ClinicAPIClient: @unchecked Sendable {
    private static let json = JSONEncoder()
    private static let decoder = JSONDecoder()
    private let session = URLSession.shared
    private let demo = DemoClinicData()
    private var settings: AppSettings

    /// Set once by `AuthController` after it constructs both objects. `weak`
    /// so the two types never form a retain cycle.
    public weak var tokenProvider: ClinicSessionTokenProviding?

    public init(settings: AppSettings) {
        self.settings = settings
    }

    /// True when no HTTPS Worker URL has been configured — the console runs on
    /// local fixture data, same as the Windows app's demo mode.
    public var isDemo: Bool {
        guard let url = URL(string: settings.apiBaseUrl), url.scheme == "https" else { return true }
        return false
    }

    public func updateSettings(_ settings: AppSettings) { self.settings = settings }

    // MARK: - Clinic dashboard

    public func getDashboard() async throws -> ClinicDashboard {
        if isDemo { return demo.dashboard() }
        return try await send("GET", "/api/clinic/dashboard")
    }

    public func publishAvailability(_ update: AvailabilityUpdate) async throws {
        if isDemo { demo.publish(update); return }
        try await sendVoid("POST", "/api/clinic/availability", body: update)
    }

    public func respond(to item: ClinicRequest, decision: DecisionPayload) async throws {
        if isDemo { try demo.decide(id: item.id, decision: decision); return }
        if item.searchTarget {
            let payload = SearchDecisionPayload(
                decision: decision.decision, responseType: decision.responseType, availableAt: decision.availableAt,
                arrivalWindowMinutes: decision.arrivalWindowMinutes, holdMinutes: decision.holdMinutes,
                waitMin: decision.waitMin, waitMax: decision.waitMax, note: decision.note
            )
            let path = "/api/clinic/search-targets/\(Self.escape(item.id))/decision"
            try await sendVoid("POST", path, body: payload)
        } else {
            let payload = IntakeDecisionPayload(
                decision: decision.decision == "offer" ? "accept" : "decline",
                arrivalWindowMinutes: decision.arrivalWindowMinutes, note: decision.note
            )
            let path = "/api/clinic/intakes/\(Self.escape(item.id))/decision"
            try await sendVoid("POST", path, body: payload)
        }
    }

    // MARK: - Session, config

    public func getSession() async throws -> SessionDescriptor? {
        let envelope: SessionEnvelope = try await send("GET", "/api/session")
        return envelope.session
    }

    /// The address actually in use, for an error to quote. A blank one is the
    /// most common cause of "could not reach Clerk" and the least visible.
    public var configuredAddress: String {
        let address = settings.apiBaseUrl.trimmingCharacters(in: .whitespacesAndNewlines)
        return address.isEmpty ? "no Worker address" : address
    }

    public func getConfig() async throws -> AppConfig {
        try await send("GET", "/api/config")
    }

    // MARK: - Calling preferences

    public func getCallPreferences() async throws -> CallPreferences {
        if isDemo { return CallPreferences(callsEnabled: true, voicePhone: nil, locationPhone: "(510) 555-0194", quietHours: nil) }
        let envelope: CallPreferencesEnvelope = try await send("GET", "/api/clinic/call-preferences")
        return envelope.preferences
    }

    public func updateCallPreferences(_ update: CallPreferencesUpdate) async throws -> CallPreferences {
        if isDemo { return CallPreferences(callsEnabled: update.callsEnabled ?? true, voicePhone: update.voicePhone, locationPhone: "(510) 555-0194") }
        let envelope: CallPreferencesEnvelope = try await send("PATCH", "/api/clinic/call-preferences", body: update)
        return envelope.preferences
    }

    // MARK: - Payouts

    /// What Tími has transferred to this clinic and what Stripe has paid out.
    ///
    /// The tenant is never a parameter: the Worker takes it from the session,
    /// so there is no shape of this call that reads another clinic's money.
    public func getPayouts() async throws -> ClinicPayouts {
        if isDemo { return demo.payouts() }
        return try await send("GET", "/api/clinic/payouts")
    }

    // MARK: - Analytics

    /// Fire-and-forget beacon to `POST /api/analytics`.
    ///
    /// Deliberately not routed through `buildRequest`: the endpoint is public
    /// and cookieless, so the beacon must carry no session token, demo header,
    /// or any other identifier. Failures are swallowed whole and never awaited
    /// — a metrics beacon that can surface an error, or hold up an intake
    /// decision, in a clinic's console has its priorities backwards.
    public func trackEvent(_ name: String, meta: [String: String]? = nil) {
        guard !isDemo, let base = try? resolvedBase(),
              let url = URL(string: "api/analytics", relativeTo: base)?.absoluteURL else { return }
        guard let body = try? Self.json.encode(AnalyticsPayload(events: [AnalyticsEvent(name: name, meta: meta)])) else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 10
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = body
        let session = self.session
        Task { _ = try? await session.data(for: request) }
    }

    // MARK: - Tenant people management

    public func getMembers() async throws -> TenantRoster {
        try await send("GET", "/api/tenant/members")
    }

    public func addMember(email: String, role: String) async throws {
        try await sendVoid("POST", "/api/tenant/members", body: AddMemberPayload(email: email, role: role))
    }

    public func changeMemberRole(clerkUserId: String, role: String) async throws {
        let path = "/api/tenant/members/\(Self.escape(clerkUserId))"
        try await sendVoid("PATCH", path, body: RolePayload(role: role))
    }

    public func removeMember(clerkUserId: String) async throws {
        let path = "/api/tenant/members/\(Self.escape(clerkUserId))"
        try await sendVoid("DELETE", path, data: nil)
    }

    public func revokeInvitation(id: String) async throws {
        let path = "/api/tenant/invitations/\(Self.escape(id))"
        try await sendVoid("DELETE", path, data: nil)
    }

    // MARK: - Transport

    private func send<Response: Decodable>(_ method: String, _ path: String) async throws -> Response {
        let data = try await requestData(method, path, data: nil)
        do { return try Self.decoder.decode(Response.self, from: data) }
        catch { throw ClinicAPIError.invalidResponse }
    }

    private func send<Response: Decodable, Body: Encodable>(_ method: String, _ path: String, body: Body) async throws -> Response {
        let data = try await requestData(method, path, data: try Self.json.encode(body))
        do { return try Self.decoder.decode(Response.self, from: data) }
        catch { throw ClinicAPIError.invalidResponse }
    }

    /// For endpoints whose success body carries nothing the caller needs — the
    /// response is still fetched and status-checked, just never decoded, so an
    /// empty (or unexpectedly shaped) success body never surfaces as an error.
    private func sendVoid(_ method: String, _ path: String, data: Data?) async throws {
        _ = try await requestData(method, path, data: data)
    }

    private func sendVoid<Body: Encodable>(_ method: String, _ path: String, body: Body) async throws {
        _ = try await requestData(method, path, data: try Self.json.encode(body))
    }

    /// Sends one request and, on a 401 while a live Clerk session exists,
    /// force-refreshes the token and retries exactly once — mirroring the
    /// Windows client's `SendAsync`. Demo headers (`x-demo-role` /
    /// `x-demo-tenant-id`) are only ever attached when there is no session
    /// AND the resolved host is loopback — never against a production HTTPS
    /// Worker, matching the same fix just made on the Windows side.
    private func requestData(_ method: String, _ path: String, data: Data?) async throws -> Data {
        let base = try resolvedBase()
        let url = URL(string: path.trimmingCharacters(in: CharacterSet(charactersIn: "/")), relativeTo: base)?.absoluteURL ?? base

        var request = try await buildRequest(url: url, method: method, data: data)
        let (firstData, firstResponse) = try await session.data(for: request)
        guard let firstHTTP = firstResponse as? HTTPURLResponse else { throw ClinicAPIError.invalidResponse }

        if firstHTTP.statusCode == 401, let tokenProvider, await tokenProvider.hasSession {
            let refreshed = try await tokenProvider.forceRefreshToken()
            request.setValue("Bearer \(refreshed)", forHTTPHeaderField: "Authorization")
            let (retryData, retryResponse) = try await session.data(for: request)
            guard let retryHTTP = retryResponse as? HTTPURLResponse else { throw ClinicAPIError.invalidResponse }
            guard (200..<300).contains(retryHTTP.statusCode) else {
                throw ClinicAPIError.server(Self.extractMessage(retryData, status: retryHTTP.statusCode))
            }
            return retryData
        }

        guard (200..<300).contains(firstHTTP.statusCode) else {
            throw ClinicAPIError.server(Self.extractMessage(firstData, status: firstHTTP.statusCode))
        }
        return firstData
    }

    private func resolvedBase() throws -> URL {
        guard let base = URL(string: settings.apiBaseUrl.isEmpty ? "" : settings.apiBaseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/") else {
            throw ClinicAPIError.invalidConfiguration
        }
        return base
    }

    private func buildRequest(url: URL, method: String, data: Data?) async throws -> URLRequest {
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.timeoutInterval = 20
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        if let tokenProvider, await tokenProvider.hasSession {
            let token = try await tokenProvider.ensureFreshToken()
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        } else if Self.isLoopback(url) {
            request.setValue("clinic", forHTTPHeaderField: "x-demo-role")
            request.setValue(settings.tenantId, forHTTPHeaderField: "x-demo-tenant-id")
        } else if !Self.isPublic(url) {
            throw ClinicAPIError.signInRequired
        }

        if let data {
            request.httpBody = data
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        return request
    }

    /// Endpoints the Worker answers to anyone, and that the console must reach
    /// *before* it can sign in.
    ///
    /// `/api/config` is where the Clerk publishable key comes from, so
    /// requiring a session to fetch it is a deadlock: no config, no Clerk host,
    /// no sign-in, no session, no config. The console reported it as "Could not
    /// read https://providers.timinow.pet/api/config — Sign in to Tími before
    /// contacting a production Worker", which reads as a Worker or a Clerk
    /// problem and is neither; the request was never sent. The Windows client
    /// never hit this only because its ClerkAuthService fetches /api/config
    /// with its own HttpClient instead of going through the gated one.
    private static func isPublic(_ url: URL) -> Bool {
        url.path.hasSuffix("/api/config")
    }

    private static func isLoopback(_ url: URL) -> Bool {
        guard let host = url.host?.lowercased() else { return false }
        return host == "localhost" || host == "127.0.0.1" || host == "::1"
    }

    /// The Worker's own words, when it sent any.
    ///
    /// This looked for `message` at the top level and for `error` as a string.
    /// The Worker sends neither: every failure is
    /// `{"error": {"code": …, "message": …}}`, with `error` an object. So the
    /// message was found by nothing and every failure in this console read
    /// "Tími API error 404." — a number, with no route, no reason, and no way
    /// to tell a missing endpoint from a missing clinic. That is how a route
    /// the desktop consoles could not reach at all survived: the server said
    /// exactly what was wrong on every attempt, and the client threw it away.
    private static func extractMessage(_ data: Data, status: Int) -> String {
        if let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            if let envelope = object["error"] as? [String: Any] {
                let message = envelope["message"] as? String ?? ""
                let code = envelope["code"] as? String ?? ""
                if !message.isEmpty { return code.isEmpty ? message : "\(message) (\(code))" }
                if !code.isEmpty { return code }
            }
            if let message = object["message"] as? String, !message.isEmpty { return message }
            if let error = object["error"] as? String, !error.isEmpty { return error }
        }
        return "Tími API error \(status)."
    }

    private static func escape(_ value: String) -> String {
        value.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? value
    }
}

// MARK: - Wire payloads (private: never leaves the public bridge surface)

/// Only the fields being changed are sent — an absent one is left alone, so
/// two administrators editing different settings do not overwrite each other.
public struct CallPreferencesUpdate: Encodable, Sendable {
    public var callsEnabled: Bool?
    public var voicePhone: String?
    public var quietHours: QuietHours?
    public init(callsEnabled: Bool? = nil, voicePhone: String? = nil, quietHours: QuietHours? = nil) {
        self.callsEnabled = callsEnabled; self.voicePhone = voicePhone; self.quietHours = quietHours
    }
}

private struct SearchDecisionPayload: Encodable {
    var decision: String
    var responseType: String
    var availableAt: String?
    var arrivalWindowMinutes: Int
    var holdMinutes: Int
    var waitMin: Int
    var waitMax: Int
    var note: String
}

private struct IntakeDecisionPayload: Encodable {
    var decision: String
    var arrivalWindowMinutes: Int
    var note: String
}

private struct AddMemberPayload: Encodable { var email: String; var role: String }
private struct RolePayload: Encodable { var role: String }

/// The console-analytics contract (POST /api/analytics): {events: [{name,
/// path?, meta?}]}, at most 25 events, and — because the endpoint is
/// cookieless by design — nothing that identifies the operator or the tenant.
private struct AnalyticsEvent: Encodable { var name: String; var meta: [String: String]? }
private struct AnalyticsPayload: Encodable { var events: [AnalyticsEvent] }
